polkadot/xcm/xcm-builder/src/nonfungible_adapter.rs

```diff
@@ -0,0 +1,319 @@
+use frame_support::{
+	ensure,
+	traits::{tokens::nonfungible, Get},
+};
+use sp_std::{marker::PhantomData, prelude::*, result};
+use xcm::latest::prelude::*;
+use xcm_executor::traits::{
+	ConvertLocation, Error as MatchError, MatchesNonFungible, TransactAsset,
+};
+
+const LOG_TARGET: &str = "xcm::nonfungible_adapter";
+
+pub struct NonFungibleTransferAdapter<Asset, Matcher, AccountIdConverter, AccountId>(
+	PhantomData<(Asset, Matcher, AccountIdConverter, AccountId)>,
+);
+impl<
+		Asset: nonfungible::Transfer<AccountId>,
+		Matcher: MatchesNonFungible<Asset::ItemId>,
+		AccountIdConverter: ConvertLocation<AccountId>,
+		AccountId: Clone, // can't get away without it since Currency is generic over it.
+	> TransactAsset for NonFungibleTransferAdapter<Asset, Matcher, AccountIdConverter, AccountId>
+{
+	fn transfer_asset(
+		what: &MultiAsset,
+		from: &MultiLocation,
+		to: &MultiLocation,
+		context: &XcmContext,
+	) -> result::Result<xcm_executor::Assets, XcmError> {
+		log::trace!(
+			target: LOG_TARGET,
+			"transfer_asset what: {:?}, from: {:?}, to: {:?}, context: {:?}",
+			what,
+			from,
+			to,
+			context,
+		);
+		// Check we handle this asset.
+		let instance = Matcher::matches_nonfungible(what).ok_or(MatchError::AssetNotHandled)?;
+		let destination = AccountIdConverter::convert_location(to)
+			.ok_or(MatchError::AccountIdConversionFailed)?;
+		Asset::transfer(&instance, &destination)
+			.map_err(|e| XcmError::FailedToTransactAsset(e.into()))?;
+		Ok(what.clone().into())
+	}
+}
+
+pub struct NonFungibleMutateAdapter<Asset, Matcher, AccountIdConverter, AccountId, CheckingAccount>(
+	PhantomData<(Asset, Matcher, AccountIdConverter, AccountId, CheckingAccount)>,
+);
+
+impl<
+		Asset: nonfungible::Mutate<AccountId>,
+		Matcher: MatchesNonFungible<Asset::ItemId>,
+		AccountIdConverter: ConvertLocation<AccountId>,
+		AccountId: Clone + Eq, // can't get away without it since Currency is generic over it.
+		CheckingAccount: Get<Option<(AccountId, MintLocation)>>,
+	> NonFungibleMutateAdapter<Asset, Matcher, AccountIdConverter, AccountId, CheckingAccount>
+{
+	fn can_accrue_checked(instance: Asset::ItemId) -> XcmResult {
+		ensure!(Asset::owner(&instance).is_none(), XcmError::NotDepositable);
+		Ok(())
+	}
+	fn can_reduce_checked(checking_account: AccountId, instance: Asset::ItemId) -> XcmResult {
+		// This is an asset whose teleports we track.
+		let owner = Asset::owner(&instance);
+		ensure!(owner == Some(checking_account), XcmError::NotWithdrawable);
+		ensure!(Asset::can_transfer(&instance), XcmError::NotWithdrawable);
+		Ok(())
+	}
+	fn accrue_checked(checking_account: AccountId, instance: Asset::ItemId) {
+		let ok = Asset::mint_into(&instance, &checking_account).is_ok();
+		debug_assert!(ok, "`mint_into` cannot generally fail; qed");
+	}
+	fn reduce_checked(instance: Asset::ItemId) {
+		let ok = Asset::burn(&instance, None).is_ok();
+		debug_assert!(ok, "`can_check_in` must have returned `true` immediately prior; qed");
+	}
```

Comment:
Perhaps these should be proper log messages and not `debug_assert`s. You are assuming that the user has e.g. called `can_check_in` prior to calling this function, but there is no guarantee. If the trait doesn't allow you to return an error here, should probably log it.

Reply:
I used `debug_assert` here to keep consistent with other adapter implementations.